docs/markdown/Tutorials/test-custom-plugin-goal.md

```diff
@@ -0,0 +1,280 @@
+
+In the following code snippet, we define a set of files to be created (in a temporary directory that Pants manages for us), the backends to be used (Python and our custom plugin), and a Pants command to be run. By reading the `stdout` of a process, we can confirm the plugin works as expected (conveniently ignoring any unrelated warnings that Pants may have produced).
+
+```python
+import json
+from pathlib import Path
+
+from pants.testutil.pants_integration_test import run_pants, setup_tmpdir
+
+build_root_marker = Path.cwd().joinpath("BUILDROOT")
+
+
+def test_reading_project_version_target() -> None:
+    """Run a full Pants process as it would run on the command line."""
+    project_files = {
+        "project/BUILD": "version_file(source='VERSION')",
+        "project/VERSION": "10.6.1",
+    }
+    # This is a limitation of the current implementation.
+    # See https://github.com/pantsbuild/pants/issues/12760.
+    build_root_marker.touch()
+    with setup_tmpdir(project_files) as tmpdir:
+        result = run_pants(
+            [
+                (
+                    "--backend-packages="
+                    "['pants.backend.python', 'internal_plugins.project_version']"
+                ),
+                "project-version",
+                "--as-json",
+                f"{tmpdir}/project:",
+            ],
+        )
+        result.assert_success()
+        assert result.stdout.strip() == json.dumps(
+            {"path": f"{tmpdir}/project/VERSION", "version": "10.6.1"}
+        )
+    build_root_marker.unlink()
+```
+
+These tests do not need any special bootstrapping and can be run just like any other tests you may have in the repository with the `test` goal. They, however, are slow, and if there are lots of test cases to check (e.g. you want to test usage of flags and targets with various fields set), it may soon become impractical to run them often enough. You would most likely want to test your plugin logic in a more isolated fashion.
+
+## Testing goal rules
+
+You can exercise the goal rule by using [`rule_runner.run_goal_rule()`](https://www.pantsbuild.org/docs/rules-api-testing#testing-goal_rules) which runs very fast and does not start a full Pants process. In the test below, we register all rules from the `project_version` plugin with the `RuleRunner` so that the engine can find them when a test is run. These tests scale nicely and if your plugins are fairly simple, they may suffice.
```

Comment:
there's a shorter form for docs links. Example: 
```
[`rule_runner.run_goal_rule()`](doc:rules-api-testing#testing-goal_rules)
```

Reply:
Oh yes, forgot about that, thank you! Fixed.